plugins/inputs/mqtt_consumer/mqtt_consumer.go

```diff
@@ -240,6 +246,16 @@ func compareTopics(expected []string, incoming []string) bool {
 }
 
 func (m *MQTTConsumer) onMessage(acc telegraf.TrackingAccumulator, msg mqtt.Message) error {
+	byteCount := unsafe.Sizeof(msg) +
+		unsafe.Sizeof(msg.Payload()) +
+		unsafe.Sizeof(msg.MessageID()) +
+		unsafe.Sizeof(msg.Duplicate()) +
+		unsafe.Sizeof(msg.Qos()) +
+		unsafe.Sizeof(msg.Topic()) +
+		unsafe.Sizeof(msg.Retained())
```

Comment:
Correct, see the go package docs: [unsafe.Sizeof](https://pkg.go.dev/unsafe#Sizeof)

>Sizeof takes an expression x of any type and returns the size in bytes of a hypothetical variable v as if v was declared via var v = x. The size does not include any memory possibly referenced by x. For instance, if x is a slice, Sizeof returns the size of the slice descriptor, not the size of the memory referenced by the slice. The return value of Sizeof is a Go constant.

```suggestion
	byteCount := unsafe.Sizeof(msg)
	byteCount += len(msg.Payload())
	byteCount += len(msg.Topic())
```

But that might be also not correct, since `msg` also contains a `sync.Once` field. So maybe don't do the `Sizeof` for the struct itself, but for every field, except for strings where you use `len`. [inspiration](https://stackoverflow.com/questions/31496804/how-to-get-the-size-of-struct-and-its-contents-in-bytes-in-golang)

Reply:
I can't decide on that, but it seems to me this (and your) approach has nothing to do with the amount of bytes received on the network, but only how much they take in memory, which is dependent on programming language, compiler optimisations and cpu architecture (eg 32 bit vs 64 bit).

Reply:
I believe what I have now should be doing it correctly, but just so I can make sure, what do you mean when you say ask this on the used mqtt package project?

Reply:
What you have now seems even worse. I mean the mqtt library used is https://github.com/eclipse/paho.mqtt.golang. So maybe look into their documentation or ask their support on how to get the actual message size retrieved by the network interface..